### PR TITLE
Use Patch instead of Update on jobframework.

### LINKED ISSUE #2269: Kueue might lose fields when admitting jobs
**What happened**:

Whenever Kueue APIs are a version behind the latest version for a particular job type, there is the risk of dropping fields, because of the use of `Update`.

We should be using `Patch` instead.

**What you expected to happen**:

Kueue to be resilient to upgrades of Job CRDs.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Kueue version (use `git describe --tags --dirty --always`):
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by mbobrovskyi**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
Use `Patch` instead of `Update` on jobframework to prevent the risk of dropping fields.

#### Which issue(s) this PR fixes:

Fixes #2269

#### Does this PR introduce a user-facing change?

```release-note
Use Patch instead of Update on jobframework to prevent the risk of dropping fields.
```

### CODE REVIEW

**Review by trasc — COMMENTED** [review]
/lgtm

**Review by mimowo — COMMENTED** [review]
LGTM, just one nit.

File: pkg/controller/jobs/job/job_controller.go [hunk 1]
```diff
@@ -162,28 +163,39 @@ func (j *Job) Suspend() {
...
 	if j.Status.StartTime != nil {
+		objectOriginal := j.Object().DeepCopyObject().(client.Object)
 		j.Status.StartTime = nil
-		if err := c.Status().Update(ctx, j.Object()); err != nil {
+		object := j.Object()
+		if err := clientutil.PatchStatus(ctx, c, objectOriginal, object); err != nil {
 			return stoppedNow, fmt.Errorf("reset status: %w", err)
 		}
+		if err := c.Get(ctx, client.ObjectKeyFromObject(object), object); err != nil {
+			return false, fmt.Errorf("get after suspend: %w", err)
```
[the remark this answers is not in the document]
Comment (mbobrovskyi) [reply pkg/controller/jobs/job/job_controller.go]: Done

[the remark this answers is not in the document]
Comment (mbobrovskyi) [reply pkg/controller/jobs/job/job_controller.go]: Oops, sorry

[the remark this answers is not in the document]
Comment (mbobrovskyi) [reply pkg/controller/jobs/job/job_controller.go]: Done

File: pkg/controller/jobs/job/job_controller.go [hunk 2]
```diff
@@ -162,28 +163,39 @@ func (j *Job) Suspend() {
 	j.Spec.Suspend = ptr.To(true)
 }
 
-func (j *Job) Stop(ctx context.Context, c client.Client, podSetsInfo []podset.PodSetInfo, _ jobframework.StopReason, eventMsg string) (bool, error) {
+func (j *Job) Stop(ctx context.Context, c client.Client, podSetsInfo []podset.PodSetInfo, _ jobframework.StopReason, _ string) (bool, error) {
 	stoppedNow := false
 	if !j.IsSuspended() {
+		objectOriginal := j.Object().DeepCopyObject().(client.Object)
```
[the remark this answers is not in the document]
Comment (mbobrovskyi) [reply pkg/controller/jobs/job/job_controller.go:169]: Unfortunately, this solution doesn't cover all cases. For example [this](https://github.com/kubernetes-sigs/kueue/blob/main/pkg/controller/jobs/job/job_controller.go#L194-L200) one.

[the remark this answers is not in the document]
Comment (mimowo) [reply pkg/controller/jobs/job/job_controller.go:169]: Ah, got it, so we would need to have `update func() (updated bool, err error)` passed?

[the remark this answers is not in the document]
Comment (mbobrovskyi) [reply pkg/controller/jobs/job/job_controller.go:169]: Fixed on #2553

**Comment by mimowo** [comment]
/reopen 
Let's keep it open to keep track of the outstanding work in https://github.com/kubernetes-sigs/kueue/pull/2553 which is not just cleanup but also semantic change from non-strict to strict patches.

**Pull request merged**